Test failure due to `float0` array in gradient computation for `BernoulliProbs` distribution in `jax==0.4.34`

Two test cases failed when testing the `test_log_prob_gradient` function for the `BernoulliProbs` distribution. The error was caused by an attempt to compute gradients w.r.t. integer arguments, resulting in generating a `float0` array. `float0` arrays are incompatible with non-trivial vector spaces and do not support operations by design. The JAX library raised a `TypeError`, indicating that `float0` arrays were used in a multiplication operation, which is unsupported. To fix this, the test needs to handle integer inputs correctly or convert `float0` arrays to a compatible dtype such as `float`.

```bash
pytest test/test_distributions.py::test_log_prob_gradient -k BernoulliProbs
```
<details><summary>Error</summary>
<p>

```bash
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.3.2, pluggy-1.5.0 -- /home/gradf/miniforge3/envs/npyroenv/bin/python3
cachedir: .pytest_cache
rootdir: /media/gradf/academia/numpyro
configfile: pyproject.toml
collecting ... collected 216 items / 214 deselected / 2 selected

test/test_distributions.py::test_log_prob_gradient[BernoulliProbs-<lambda>-params153] FAILED [ 50%]
test/test_distributions.py::test_log_prob_gradient[BernoulliProbs-<lambda>-params154] FAILED [100%]

=================================== FAILURES ===================================
__________ test_log_prob_gradient[BernoulliProbs-<lambda>-params153] ___________

jax_dist = <class 'numpyro.distributions.discrete.BernoulliProbs'>
sp_dist = <function <lambda> at 0x7fc9c453ab60>, params = (0.2,)

    @pytest.mark.parametrize(
        "jax_dist, sp_dist, params", CONTINUOUS + DISCRETE + DIRECTIONAL
    )
    def test_log_prob_gradient(jax_dist, sp_dist, params):
        if jax_dist in [dist.LKJ, dist.LKJCholesky]:
            pytest.skip("we have separated tests for LKJCholesky distribution")
        if jax_dist is _ImproperWrapper:
            pytest.skip("no param for ImproperUniform to test for log_prob gradient")
        if (
            jax_dist in [dist.DoublyTruncatedPowerLaw]
            and jnp.result_type(float) == jnp.float32
        ):
            pytest.skip("DoublyTruncatedPowerLaw is tested with x64 only.")
    
        rng_key = random.PRNGKey(0)
        value = jax_dist(*params).sample(rng_key)
    
        def fn(*args):
            return jnp.sum(jax_dist(*args).log_prob(value))
    
        eps = 1e-3
        for i in range(len(params)):
            if jax_dist is dist.EulerMaruyama and i == 1:
                # skip taking grad w.r.t. sde_fn
                continue
            if jax_dist is _SparseCAR and i == 3:
                # skip taking grad w.r.t. adj_matrix
                continue
            if jax_dist is dist.ZeroSumNormal and i != 0:
                # skip taking grad w.r.t. event_shape
                continue
            if isinstance(
                params[i], dist.Distribution
            ):  # skip taking grad w.r.t. base_dist
                continue
            if jax_dist is dist.DoublyTruncatedPowerLaw and i != 0:
                continue
            if params[i] is None or jnp.result_type(params[i]) in (jnp.int32, jnp.int64):
                continue
>           actual_grad = jax.grad(fn, i)(*params)

test/test_distributions.py:1872: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_distributions.py:1851: in fn
    return jnp.sum(jax_dist(*args).log_prob(value))
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/numpyro/distributions/util.py:706: in wrapper
    log_prob = log_prob_fn(self, *args, **kwargs)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/numpyro/distributions/discrete.py:92: in log_prob
    return xlogy(value, ps_clamped) + xlog1py(1 - value, -ps_clamped)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/scipy/special.py:504: in _xlogy_jvp
    return result, (x_dot * lax.log(y) + y_dot * x / y).astype(result.dtype)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:573: in deferring_binary_op
    return binary_op(*args)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/ufunc_api.py:177: in __call__
    return call(*args)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/ufuncs.py:1142: in _multiply
    x, y = promote_args("multiply", x, y)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/util.py:355: in promote_args
    _check_no_float0s(fun_name, *args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fun_name = 'multiply'
args = (Traced<ShapedArray(float0[])>with<DynamicJaxprTrace(level=3/0)>, Traced<ShapedArray(float32[])>with<DynamicJaxprTrace(level=3/0)>)

    def check_no_float0s(fun_name: str, *args: Any):
      """Check if none of the args have dtype float0."""
      if any(dtypes.dtype(arg) == dtypes.float0 for arg in args):
>       raise TypeError(
            f"Called {fun_name} with a float0 array. "
            "float0s do not support any operations by design because they "
            "are not compatible with non-trivial vector spaces. No implicit dtype "
            "conversion is done. You can use np.zeros_like(arr, dtype=np.float) "
            "to cast a float0 array to a regular zeros array. \n"
            "If you didn't expect to get a float0 you might have accidentally "
            "taken a gradient with respect to an integer argument.")
E       TypeError: Called multiply with a float0 array. float0s do not support any operations by design because they are not compatible with non-trivial vector spaces. No implicit dtype conversion is done. You can use np.zeros_like(arr, dtype=np.float) to cast a float0 array to a regular zeros array. 
E       If you didn't expect to get a float0 you might have accidentally taken a gradient with respect to an integer argument.
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/util.py:326: TypeError
__________ test_log_prob_gradient[BernoulliProbs-<lambda>-params154] ___________

jax_dist = <class 'numpyro.distributions.discrete.BernoulliProbs'>
sp_dist = <function <lambda> at 0x7fc9c453ab60>, params = (array([0.2, 0.7]),)

    @pytest.mark.parametrize(
        "jax_dist, sp_dist, params", CONTINUOUS + DISCRETE + DIRECTIONAL
    )
    def test_log_prob_gradient(jax_dist, sp_dist, params):
        if jax_dist in [dist.LKJ, dist.LKJCholesky]:
            pytest.skip("we have separated tests for LKJCholesky distribution")
        if jax_dist is _ImproperWrapper:
            pytest.skip("no param for ImproperUniform to test for log_prob gradient")
        if (
            jax_dist in [dist.DoublyTruncatedPowerLaw]
            and jnp.result_type(float) == jnp.float32
        ):
            pytest.skip("DoublyTruncatedPowerLaw is tested with x64 only.")
    
        rng_key = random.PRNGKey(0)
        value = jax_dist(*params).sample(rng_key)
    
        def fn(*args):
            return jnp.sum(jax_dist(*args).log_prob(value))
    
        eps = 1e-3
        for i in range(len(params)):
            if jax_dist is dist.EulerMaruyama and i == 1:
                # skip taking grad w.r.t. sde_fn
                continue
            if jax_dist is _SparseCAR and i == 3:
                # skip taking grad w.r.t. adj_matrix
                continue
            if jax_dist is dist.ZeroSumNormal and i != 0:
                # skip taking grad w.r.t. event_shape
                continue
            if isinstance(
                params[i], dist.Distribution
            ):  # skip taking grad w.r.t. base_dist
                continue
            if jax_dist is dist.DoublyTruncatedPowerLaw and i != 0:
                continue
            if params[i] is None or jnp.result_type(params[i]) in (jnp.int32, jnp.int64):
                continue
>           actual_grad = jax.grad(fn, i)(*params)

test/test_distributions.py:1872: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test/test_distributions.py:1851: in fn
    return jnp.sum(jax_dist(*args).log_prob(value))
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/numpyro/distributions/util.py:706: in wrapper
    log_prob = log_prob_fn(self, *args, **kwargs)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/numpyro/distributions/discrete.py:92: in log_prob
    return xlogy(value, ps_clamped) + xlog1py(1 - value, -ps_clamped)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/scipy/special.py:504: in _xlogy_jvp
    return result, (x_dot * lax.log(y) + y_dot * x / y).astype(result.dtype)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:573: in deferring_binary_op
    return binary_op(*args)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/ufunc_api.py:177: in __call__
    return call(*args)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/ufuncs.py:1142: in _multiply
    x, y = promote_args("multiply", x, y)
/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/util.py:355: in promote_args
    _check_no_float0s(fun_name, *args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

fun_name = 'multiply'
args = (Traced<ShapedArray(float0[2])>with<DynamicJaxprTrace(level=3/0)>, Traced<ShapedArray(float32[2])>with<DynamicJaxprTrace(level=3/0)>)

    def check_no_float0s(fun_name: str, *args: Any):
      """Check if none of the args have dtype float0."""
      if any(dtypes.dtype(arg) == dtypes.float0 for arg in args):
>       raise TypeError(
            f"Called {fun_name} with a float0 array. "
            "float0s do not support any operations by design because they "
            "are not compatible with non-trivial vector spaces. No implicit dtype "
            "conversion is done. You can use np.zeros_like(arr, dtype=np.float) "
            "to cast a float0 array to a regular zeros array. \n"
            "If you didn't expect to get a float0 you might have accidentally "
            "taken a gradient with respect to an integer argument.")
E       TypeError: Called multiply with a float0 array. float0s do not support any operations by design because they are not compatible with non-trivial vector spaces. No implicit dtype conversion is done. You can use np.zeros_like(arr, dtype=np.float) to cast a float0 array to a regular zeros array. 
E       If you didn't expect to get a float0 you might have accidentally taken a gradient with respect to an integer argument.
E       --------------------
E       For simplicity, JAX has removed its internal frames from the traceback of the following exception. Set JAX_TRACEBACK_FILTERING=off to include these.

/home/gradf/miniforge3/envs/npyroenv/lib/python3.11/site-packages/jax/_src/numpy/util.py:326: TypeError
=========================== short test summary info ============================
FAILED test/test_distributions.py::test_log_prob_gradient[BernoulliProbs-<lambda>-params153]
FAILED test/test_distributions.py::test_log_prob_gradient[BernoulliProbs-<lambda>-params154]
====================== 2 failed, 214 deselected in 2.41s =======================
```

</p>
</details>

## Info

```bash
jax              0.4.34
jaxlib           0.4.34
numpy            2.1.2
numpyro          0.15.3
```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -14,7 +14,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.10]
+        python-version: ["3.10"]
 
     steps:
     - uses: actions/checkout@v2
@@ -50,7 +50,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.10]
+        python-version: ["3.10"]
 
     steps:
     - uses: actions/checkout@v2
@@ -83,7 +83,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.10]
+        python-version: ["3.10"]
 
     steps:
     - uses: actions/checkout@v2
@@ -126,7 +126,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.10]
+        python-version: ["3.10"]
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -14,7 +14,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.9]
+        python-version: [3.10]
 
     steps:
     - uses: actions/checkout@v2
@@ -50,7 +50,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.9]
+        python-version: [3.10]
 
     steps:
     - uses: actions/checkout@v2
@@ -83,7 +83,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.9]
+        python-version: [3.10]
 
     steps:
     - uses: actions/checkout@v2
@@ -126,7 +126,7 @@ jobs:
     needs: lint
     strategy:
       matrix:
-        python-version: [3.9]
+        python-version: [3.10]
 
     steps:
     - uses: actions/checkout@v2

--- a/numpyro/distributions/discrete.py
+++ b/numpyro/distributions/discrete.py
@@ -89,7 +89,8 @@ class BernoulliProbs(Distribution):
     @validate_sample
     def log_prob(self, value):
         ps_clamped = clamp_probs(self.probs)
-        return xlogy(value, ps_clamped) + xlog1py(1 - value, -ps_clamped)
+        value = jnp.array(value, jnp.result_type(float))
+        return xlogy(value, ps_clamped) + xlog1py(1.0 - value, -ps_clamped)
 
     @lazy_property
     def logits(self):
@@ -190,6 +191,7 @@ class BinomialProbs(Distribution):
 
     @validate_sample
     def log_prob(self, value):
+        value = jnp.array(value, jnp.result_type(float))
         log_factorial_n = gammaln(self.total_count + 1)
         log_factorial_k = gammaln(value + 1)
         log_factorial_nmk = gammaln(self.total_count - value + 1)
@@ -558,6 +560,7 @@ class MultinomialProbs(Distribution):
     def log_prob(self, value):
         if self._validate_args:
             self._validate_sample(value)
+        value = jnp.array(value, jnp.result_type(float))
         return gammaln(self.total_count + 1) + jnp.sum(
             xlogy(value, self.probs) - gammaln(value + 1), axis=-1
         )

--- a/numpyro/distributions/discrete.py
+++ b/numpyro/distributions/discrete.py
@@ -90,7 +90,7 @@ class BernoulliProbs(Distribution):
     def log_prob(self, value):
         ps_clamped = clamp_probs(self.probs)
         value = jnp.array(value, jnp.result_type(float))
-        return xlogy(value, ps_clamped) + xlog1py(1.0 - value, -ps_clamped)
+        return xlogy(value, ps_clamped) + xlog1py(1 - value, -ps_clamped)
 
     @lazy_property
     def logits(self):
@@ -263,12 +263,13 @@ class BinomialLogits(Distribution):
 
     @validate_sample
     def log_prob(self, value):
-        log_factorial_n = gammaln(self.total_count + 1)
+        total_count = jnp.array(self.total_count, dtype=jnp.result_type(float))
+        log_factorial_n = gammaln(total_count + 1)
         log_factorial_k = gammaln(value + 1)
-        log_factorial_nmk = gammaln(self.total_count - value + 1)
+        log_factorial_nmk = gammaln(total_count - value + 1)
         normalize_term = (
             self.total_count * jnp.clip(self.logits, 0)
-            + xlog1py(self.total_count, jnp.exp(-jnp.abs(self.logits)))
+            + xlog1py(total_count, jnp.exp(-jnp.abs(self.logits)))
             - log_factorial_n
         )
         return (
@@ -558,8 +559,6 @@ class MultinomialProbs(Distribution):
 
     @validate_sample
     def log_prob(self, value):
-        if self._validate_args:
-            self._validate_sample(value)
         value = jnp.array(value, jnp.result_type(float))
         return gammaln(self.total_count + 1) + jnp.sum(
             xlogy(value, self.probs) - gammaln(value + 1), axis=-1

